### ISSUE TITLE
Math Node, VMath, Trigonometry, and Logic

### ISSUE THREAD

**Issue by ly29** [Sverchok/Sverchok#12]
Following the mostly successful multi modal node experiment in #11 I now turn to the basic node set starting with math nodes.
In the old Sverchok I think perhaps math node got to many modes, there are 44 and the menu is somewhat hard to navigate. 
I propose to split up trigonometry into a different node directly since it is clear divide to make is somewhat more resonable, for now I will not add all functions directly...

See https://github.com/Sverchok/Sverchok/pull/25 for the current pullrequest for Math node. And modes which are fully or partially implemented
See https://github.com/Sverchok/Sverchok/pull/28 for Vector Math node branch / PR.

**Comment by zeffii** [comment]
finally :)

**Comment by ly29** [comment]
This is also a good time to face the issue of making arrays broadcastable as the numpy term for list matching is.
https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html
```python
def is_broadcastable(a, b):
    return all((x == 1 or y == 1 or x == y) for x, y in zip(a.shape[::-1], b.shape[::-1]))
```

**Comment by zeffii** [comment]
for the Trig Node
```
-  SINE
-  COSINE
-  TANGENT
-  ARCSINE
-  ARCCOSINE
-  ARCTANGENT
-  DEGREES
-  RADIANS
-  ACOSH
-  ASINH
-  ATANH
-  COSH
-  SINH
-  TANH
-  PI * N, where N defaults to 1.0
-  PHI * N
-  TAU * N
+
-  SIN(x), COS(x)

```

**Comment by ly29** [comment]
@zeffii 
Feel up for it? Do it without considering broadcasting then we can apply that knowledge.

**Comment by zeffii** [comment]
yeah i'll do this.

**Comment by ly29** [comment]
@zeffii 
There is bug with many outputs, chasing it down so if you have trouble with that don't worry

**Comment by zeffii** [comment]
just doing the boilerplate atm.. haven't tried running it yet :)

**Comment by ly29** [comment]
Well now it should work anyway.

**Comment by ly29** [comment]
Also, remember that you can expect a ```np.array``` object so: ```np.sin(x)``` instead of ```sin(x)```

**Comment by ly29** [comment]
Logic node
```python
    mode_items = [
        ("AND",             "And",          "", 1),
        ("OR",              "Or",           "", 2),
        ("IF",              "If",           "", 3),
        ("NOT",             "Not",          "", 4),
        ("NAND",            "Nand",         "", 5),
        ("NOR",             "Nor",          "", 6),
        ("XOR",             "Xor",          "", 7),
        ("XNOR",            "Xnor",         "", 8),
        ("LESS",             "<",           "", 9),
        ("BIG",              ">",           "", 10),
        ("EQUAL",            "==",          "", 11),
        ("NOT_EQ",           "!=",          "", 12),
        ("LESS_EQ",          "<=",          "", 13),
        ("BIG_EQ",           ">=",          "", 14),
        ("TRUE",             "True",        "", 15),
        ("FALSE",            "False",       "", 16),
        ("ISCLOSE",            "Isclose",       "", 16),

        ]
```

**Comment by zeffii** [comment]
add https://docs.python.org/3/library/math.html#math.isclose ? / numpy equivalent https://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.isclose.html

**Comment by ly29** [comment]
isclose makes so much more sense than ```==``` for most our things

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/22439563/6836521e-e730-11e6-9d14-71a8557a71ac.png)

**Comment by ly29** [comment]
Nice

**Comment by zeffii** [comment]
i have this fantasy where the Number slider has a property/bool that lets us switch to int / float on the fly

**Comment by ly29** [comment]
I know and I agree.

**Comment by zeffii** [comment]
this isn't clear to me,
```
@node_func(bl_idname=local_name, id=62)
@constant_times_n
def phi(n: Number = 1.0) -> Number:
    return np.empty((1 + 5 ** 0.5) / 2) * n


``` 
is producing an extra `[  ]` around the value

**Comment by zeffii** [comment]
also...  `math` could do with `copy sign`  ,   copy the sign of x and return the values of y signed that way.

**Comment by ly29** [comment]
That is point of the constant decorator. Maybe a bit decorator happy lately.

All value have to be array for some functions to work.

**Comment by ly29** [comment]
```python
>>> import numpy as np
>>> np.array(5)
array(5)

>>> len(np.array(5))
Traceback (most recent call last):
  File "<blender_console>", line 1, in <module>
TypeError: len() of unsized object
```

**Comment by zeffii** [comment]
ahhh i see!

**Comment by ly29** [comment]
There are some assumptions about how things work that I have to explain to make things clear.

**Comment by zeffii** [comment]
and because the decorator also wraps this into an np.array i don't need to do np.empty(constant) ?

**Comment by ly29** [comment]
It should be fine I think

**Comment by ly29** [comment]
Regarding switchable types I want to do it in a clean and extensible way, I will do soon but holding until a clean solution comes to mine. I think once I formalize some things it will become clear.

Actually the constant times function don't need the constant wrapper since n will be of type ```np.array``` with ```shape=(x,)```

**Comment by ly29** [comment]
Some functions to make broadcasting possible. I think same shape and 1+ many are the common cases so the uncommon being a bit slower but working makes sense.

```python
@compat
def fx(a, b):
    return a+b

>>> fx(np.array([2,3]), np.linspace(0,1, 5))
array([ 2.  ,  3.25,  3.5 ,  3.75,  4.  ])
```

```python
def compat(func):
    @wraps(func)
    def inner(a, b):
        return func(*make_compatible(a, b))
    return inner

def make_compatible(a, b):
    if is_broadcastable(a, b):
        return a, b
    shape = tuple(max(x, y) for x, y in zip(a.shape, b.shape))
    return array_as(a, shape), array_as(b, shape)

def array_as(a, shape):
    if a.shape == shape:
        return a
    new_a = np.empty(shape, dtype=a.dtype)
    new_a[:len(a)] = a
    new_a[len(a):] = a[-1]
    return new_a


def is_broadcastable(a, b):
    return all((x == 1 or y == 1 or x == y) for x, y in zip(a.shape[::-1], b.shape[::-1]))
```

**Comment by ly29** [comment]
Made a gist and tested the overhead for adding a 1 + 1000 numpy array, the results could be better.
[timings](https://gist.github.com/ly29/6e6f7a4bb895def2077d25b919be1da2.js)

```python
 >>> print(min(timeit.Timer("fx(x,y)",setup=setup).repeat(7, 1000)))
0.0020394666666589956

>>> print(min(timeit.Timer("cfx(x,y)",setup=setup).repeat(7, 1000)))
0.003827626666662809
```

For 10000 + 1
```python
>>> print(min(timeit.Timer("fx(x,y)",setup=setup).repeat(7, 1000)))
0.006587306666688164

>>> print(min(timeit.Timer("cfx(x,y)",setup=setup).repeat(7, 1000)))
0.006900480000012976
```

The good thing is that the overhead is mostly constant and not a factor of N...

Is it worth it? I think so.
With linspace count to 100000.

```python
# add x to y
>>> print(min(timeit.Timer("fx(x,y)",setup=setup).repeat(7, 1000)))
0.0460450133333552

>>> print(min(timeit.Timer("cfx(x,y)",setup=setup).repeat(7, 1000)))
0.048297813333306294

# add x to x 
>>> print(min(timeit.Timer("fx(x,x)",setup=setup).repeat(7, 1000)))
0.04570069333340143

>>> print(min(timeit.Timer("cfx(x,x)",setup=setup).repeat(7, 1000)))
0.047335680000060165
```

**Comment by ly29** [comment]
The overhead can be reduced also by optimizing the away call to ```is_broadcastable```  and replacing it with something less generic but dealing with a specific situation. Anyway to early to worry about this part.

**Comment by ly29** [comment]
@zeffii 
Based on #11
So made a custom class that should be able to support logic, math, and trigonometry
```python
@node_func(bl_idname='SvRxNodeMath', multi_label="Math", id=0, cls_bases=(NodeMathBase,))
def add(x: Number = 0.0, y: Number = 0.0) -> Number:
    return x + y
```

And in the ```NodeMathBase``` Node class there is now support for the following. Right now only int and float but could be other socket types like uint, factor etc.

![custom_type](https://cloud.githubusercontent.com/assets/6241382/22499163/61f6be74-e85d-11e6-868f-0e96500b230f.PNG)

**Comment by zeffii** [comment]
also.. `reciprocal`  `1/x`
which has a pleasantly [surprising google images result](https://www.google.nl/search?q=reciprocal&source=lnms&tbm=isch&sa=X&ved=0ahUKEwja3dD7xu_RAhWM7xQKHd_VDsAQ_AUICCgB&biw=1571&bih=946#tbm=isch&q=reciprocal+structure)

**Comment by ly29** [comment]
Yeah, need to get back into the math node enough to make a list of missing modes.

**Comment by ly29** [comment]
Considering what we discussed in #17 about how the ```node_func``` gets to heavy and that you shouldn't have to repeat your self in code. I think the following would be resonable. My removing the ```bl_idname``` it gets much visually lighter. The main problem is that the ```@node_func``` over powers the signature. @zeffii thoughts?

It works because the last used the ```bl_idname``` of the multi function is remembered.

```python
@node_func(bl_idname='SvRxNodeMath', multi_label="Math", id=0, cls_bases=(NodeMathBase,))
def add(x: Number = 0.0, y: Number = 0.0) -> Number:
    return x + y

@node_func(id=1)
def sub(x: Number = 0.0, y: Number = 0.0) -> Number:
    return x - y

@node_func(id=2)
def mul(x: Number = 0.0, y: Number = 0.0) -> Number:
    return x * y

@node_func(id=3)
def div(x: Number = 0.0, y: Number = 0.0) -> Number:
    return x / y

@node_func(id=4)
def sqrt(x: Number = 1.0) -> Number:
    return x ** .5
``` 
Compare to

```python
@node_func(bl_idname=local_name, multi_label="Trig", id=0, cls_bases=(NodeMathBase,))
def sine(x: Number = 0.0) -> Number:
    return np.sin(x)


@node_func(bl_idname=local_name, id=1)
def cosine(x: Number = 0.0) -> Number:
    return np.cos(x)



@node_func(bl_idname=local_name, id=2)
def sincos(x: Number = 0.0) -> (Number("sin"), Number("cos")):
    return np.sin(x), np.cos(x)



@node_func(bl_idname=local_name, id=3)
def degrees(x: Number = 0.0) -> Number:
    return np.degrees(x)


@node_func(bl_idname=local_name, id=4)
def radians(x: Number = 0.0) -> Number:
    return np.radians(x)
```

**Comment by zeffii** [comment]
Am obviously in favour of the lighter flavour. looks Fine .
not sitting at my dev machine atm and other stuff requires my attention.

**Comment by zeffii** [comment]
not a lot of time for this today, but does it make sense do to a shape test to test if an np.array is n*4 or 3 first ?

**Comment by ly29** [comment]
The incoming should be n,4 but right it isn't if it comes from the prop.

So it does make sense right now.

**Comment by zeffii** [comment]
unfortunately i have to go again, maybe later.

**Comment by zeffii** [comment]
OK back. but for how long. i've pushed changes / additions to both vmath and math nodes. going forward they'll be separate branches for pullrequest.

**Comment by ly29** [comment]
Ok sounds good. Will check it out. 

Going forward I focus on systematic issues once #2 is done. 

There is a bunch of things that can be easily picked of and will make it feel like a complete system.

But will be a day or two...

**Comment by zeffii** [comment]
Going through the existing modes atm.

**Comment by zeffii** [comment]
btw, sending a float to Circle's Vertex count poses a small problem
 - could perhaps indicate somehow that the input needs a `math/as_int` first?
 - auto inject `math/as int`
 - auto cast incoming floats to int? (my preference, but .. takes away information from the user)

**Comment by ly29** [comment]
I think it calls for a auto conversion, maybe it is a good idea to make a node that converts into int pop up for that case.

**Comment by zeffii** [comment]
sounds good

**Comment by zeffii** [comment]
AttributeError: 'SvVertices' object has no attribute 'shape' when connecting Vertices to VMath input

![image](https://cloud.githubusercontent.com/assets/619340/22793497/4c028988-eef0-11e6-8a3c-8d3c159b4031.png)

**Comment by ly29** [comment]
Yeah, that is slight problem, for edges and polygons we don't really have any nodes that does work on them except create output. however for the vertices I guess we need to return the raw vertices data.

**Comment by ly29** [comment]
```python
    @property
    def as_rxdata(self):
        return self.vertices.vertices, self.edges, self.faces
```

**Comment by ly29** [comment]
This feel slightly wrong and should be of some concern, it is worth giving those classes another pass and think about a better representation

**Comment by zeffii** [comment]
i thought it would be simpler to introspect from the node doing the _affecting_. and transform the incoming data into a usable representation.

I don't really have strong opinions due to not being super involved in that code ..

**Comment by ly29** [comment]
The class doesn't do anything but convert data into a ```numpy``` format, this is a sign that either should be some wrapper functions or inherit ```numpy.array```

**Comment by ly29** [comment]
https://docs.scipy.org/doc/numpy/user/basics.subclassing.html
